blueprint/core/blueprint_View.cpp

```diff
@@ -202,10 +206,123 @@ namespace blueprint
 
         if (props.contains(backgroundColorProp))
         {
-            juce::Colour c = juce::Colour::fromString(props[backgroundColorProp].toString());
-
-            if (!c.isTransparent())
-                g.fillAll(c);
+            auto colorString = props[backgroundColorProp].toString();
+            if(colorString.contains("linear-gradient")) {
```

Comment:
Similarly here, my hunch is that we should address linear-gradient as much as possible in javascript. I think perhaps we should define a minimal native API for dealing in color gradients, and then tackle `linear-gradient` as a function which maps to this minimal API.

For example, maybe our native API is this:

If the property value is a color type, we expect
* If the value type is a string, then it is of the form `ffaabbcc` and we can construct a `juce::Colour` directly
* If the value type is a dynamic object (juce::var::isDynamicObject()), then it is of the form
```
{
  type: 'linear',
  startSide: 'left',
  angle: 0.139, // perhaps we define that this number is always radians
  colourStops: [
    'ffaabbcc', // These are simply strings of the expected juce hex format
    'ffbbaacc',
  ]
}
```

Then native's only business is "is the value type string or object?" and "how do we go from string -> color (easy), and how do we go from object->colourGradient (harder, but you've done all the hard work here already)"

Reply:
ahh I got you. Yes I like the cleanliness of dynamic objects. The JUCE ColourGradient is relatively simple class, provide two colors and a pair of coordinates for each, so maybe we leave the native API as simple as that? As far as 3+ colors we can use the  `addColour()` method which only takes a relative position along the gradient line. So maybe the object we pass to native is:

```
{
  colour1: {
    color: ffaabbcc
    x: 0
    y: 0
  },
  color2: {
    color: ffaabbcc
    x: 100
    y: 100
  }
  additionalColors: [
    {
      color: ffaabbcc
      position: 0.2
    },
    {
      color: ffaabbcc
      position: 0.8
    }
  ]
}
```
We can do all logic in JS then. The only tricky thing is getting the bounds of the component in JS for the calculations before rendering. Do we have a way to do this?

Reply:
Yea good question. So, I think generally for a linear gradient we need to know two things: the angle of the gradient axis, and where the color stops are. I wonder... can we arrive at something like this for the native API?

```js
{
    angle: 1.923, // radians
    colours: [
      { hex: 'ffaabbcc', position: 0.0 },
      { hex: 'ffaabbdd', position: 0.6 },
      { hex: 'ffaaccdd', position: 1.0 },
      // ...etc
    ]
}
```

Here's my thinking: `angle` and `startSide` in the web API generally tell us how to orient this axis. But, interestingly, `startSide: left, angle: 0` is equivalent to `startSide: right, angle: 3.14`. So, I wonder, can we map _all_ notions of the gradient axis onto a single line rotated some angle from the origin?

I think we can probably map this whole thing onto something of a "unit line" where we let 0.0 denote the start of the gradient axis and 1.0 denote the end of the gradient axis, then use the angle to spin that around as needed. Now, all `position` values on each color are relative points along that unit line: 0.25 is 25% along the line, 0.8 is 80%, etc. This way, once we get into native land we can simply do (pseudocode to follow)

* Let x0 = getWidth();
* Let y0 = getHeight() / 2;
* Let x1 = 0;
* Let y1 = y0;
* Let axisStartPoint = rotate({x0, y0}, angle);
* Let axisEndPoinit = rotate({x1, y1}, angle);
* Let gradient = juce::ColourGradient(colors[0]['hex'], x0, y0, colors[1]['hex'], x1, y1));
* For each color in {colors[2]...colors[n]} gradient.addColour(colors[i]['hex'], colors[i]['position']['x;] * (x1 - x0), colors[i]['position']['x;] * (y1 - y0))

That's very crude pseudocode, but hopefully it gets the point across :D

What do you think? This let's us minimize the native API while also keeping everything abstract/relative until we get into native, at which point measuring the component is trivial

Reply:
> I think we can probably map this whole thing onto something of a "unit line" where we let 0.0 denote the start of the gradient axis and 1.0 denote the end of the gradient axis, then use the angle to spin that around as needed. Now, all `position` values on each color are relative points along that unit line: 0.25 is 25% along the line, 0.8 is 80%, etc. This way, once we get into native land we can simply do (pseudocode to follow)
> 

Yeah I think you're right, we could make it that simple. 
I know for sure that'll get us at least the basic linear gradient combinations (angles & directional words).

> * Let x0 = getWidth();
> * Let y0 = getHeight() / 2;
> * Let x1 = 0;
> * Let y1 = y0;
> * Let axisStartPoint = rotate({x0, y0}, angle);
> * Let axisEndPoinit = rotate({x1, y1}, angle);
> * Let gradient = juce::ColourGradient(colors[0]['hex'], x0, y0, colors[1]['hex'], x1, y1));
> * For each color in {colors[2]...colors[n]} gradient.addColour(colors[i]['hex'], colors[i]['position']['x;] * (x1 - x0), colors[i]['position']['x;] * (y1 - y0))
> 
Ah yup think I see your point (no pun intended).  Only change would be gradient.addColour only takes a percentage along the gradient line not coordinates. So JUCE should actually do calculation work for us there. I'm assuming the first color/point added will always be considered the zero position, but will have a play to confirm this. 

Anyway sounds like a plan, i'll start looking into it tomorrow. I think this design should actually simplify the calculation logic too.

Reply:
> Only change would be gradient.addColour only takes a percentage along the gradient line not coordinates. So JUCE should actually do calculation work for us there. I'm assuming the first color/point added will always be considered the zero position, but will have a play to confirm this.

Ah, right, yea that's convenient actually, since then we can basically take the relative position verbatim and just scale it up to percentage.

Yea, great! Thanks for tackling it :D